python-bareos/bareos/bsock/lowlevel.py

```diff
@@ -39,42 +42,77 @@ class LowLevel(object):
     Low Level socket methods to communicate with the bareos-director.
     """
 
+    @staticmethod
+    def argparser_get_bareos_parameter(args):
+        """
+        This method is usally used together with the method argparser_add_default_command_line_arguments.
+        
+        @param args: Arguments retrieved by ArgumentParser.parse_args()
+        @type args:  ArgParser.Namespace
+        
+        @return: returns the relevant parameter from args to initialize a connection.
+        @rtype: dict
+        """
+        result = {}
+        for key,value in vars(args).items():
+            if value is not None:
+                if key.startswith('BAREOS_'):
+                    bareoskey=key.split('BAREOS_', 1)[1]
+                    result[bareoskey]=value
+        return result
+
+
     def __init__(self):
         self.logger = logging.getLogger()
         self.logger.debug("init")
         self.status = None
         self.address = None
         self.password = None
+        self.pam_username = None
+        self.pam_password = None
         self.port = None
         self.dirname = None
         self.socket = None
         self.auth_credentials_valid = False
+        self.max_reconnects = 0
         self.tls_psk_enable = True
         self.tls_psk_require = False
         self.connection_type = None
+        self.protocolversion = ProtocolVersions.last
+        self.protocol_messages = ProtocolMessages()
         # identity_prefix have to be set in each class
         self.identity_prefix = u'R_NONE'
         self.receive_buffer = b''
 
+    def __del__(self):
+        self.close()
 
-    def connect(self, address, port, dirname, type, name = None, password = None):
+    def connect(self, address, port, dirname, connection_type, name = None, password = None):
         self.address = address
         self.port = int(port)
         if dirname:
             self.dirname = dirname
         else:
             self.dirname = address
-        self.connection_type = type
+        self.connection_type = connection_type
         self.name = name
-        self.password = password
+        if isinstance(password, Password):
+            self.password = password
+        else:
+            self.password = Password(password)
 
         return self.__connect()
 
 
     def __connect(self):
         connected = False
-        if self.tls_psk_require and not self.is_tls_psk_available():
-            raise bareos.exceptions.ConnectionError(u'TLS-PSK is required, but sslpsk module not loaded/available.')
+        connected_plain = False
+        auth = False
+        if self.tls_psk_require:
+            if not self.is_tls_psk_available():
+                raise bareos.exceptions.ConnectionError(u'TLS-PSK is required, but sslpsk module not loaded/available.')
+            if not self.tls_psk_enable:
```

Comment:
Thanks for pointing this out. I considered this. However, in a default installation of python-bareos  TLS-PSK is not available as the Python module sslpsk required for this functionality is not available as package in the current Linux distributions and the Bareos project decided not to ship the module in this own repositories.. While it can be installed using pip, in my environment the module has only been usable with Python 2, not with Python 3 (library problem with openssl).
So considering this, I assume that many persons will use python-bareos without TLS-PSK, maybe even in compat (protocol <= 18.2.4) mode.
Therefore in kept the possibility to disable TLS-PSK on class initiation. By this, the already implemented systemtests can control if they want to test with TLS-PSK enabled or disabled, even if the sslpsk module is installed.
However, as I agree that most people do not want to manually disable TLS-PSK, I removed the command line option --no-tls-psk in an extra commit.